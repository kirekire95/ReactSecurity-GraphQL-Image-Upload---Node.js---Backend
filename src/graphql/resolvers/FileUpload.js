import { createWriteStream } from "fs"
import path from "path"

import { generateRandomString } from "../../utilities/generateRandomString.js"

const Query = {
  uploads: () => File
}

const Mutation = {
  uploadFile: async (parent, args) => {
    const { file } = args
    // Don't need these two unless we want to store it in a database
    // const { mimetype, encoding } = await file
    const { createReadStream, filename } = await file

    const { ext } = path.parse(filename)
    const randomName = generateRandomString(12) + ext

    const stream = createReadStream()
    const pathName = path.join(__dirname, `/public/images/${randomName}`)

    await stream.pipe(createWriteStream(pathName))

    return {
      url: `https://reactsecurity---backend.herokuapp.com/images/${randomName}`
    }
  }
}

export const fileUploadResolvers = {
  Query,
  Mutation
}
