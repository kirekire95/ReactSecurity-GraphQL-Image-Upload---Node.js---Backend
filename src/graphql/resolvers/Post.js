import { AuthenticationError, UserInputError } from "apollo-server-express"

import { PostModel } from "../../models/Post.js"
import checkAuth from "../../helper/context/index.js"

const Query = {
  getPosts: async () => {
    try {
      const allPosts = await PostModel.find().sort({
        title: 1,
        description: 1,
        category: 1
      })
      return allPosts
    } catch (err) {
      throw new Error(err)
    }
  },
  getPost: async (parent, post, verifyJWT) => {
    const authenticatedUser = checkAuth(verifyJWT)
    console.log("authenticatedUser", authenticatedUser)
    try {
      const onePost = await PostModel.findById(post.postId)
      if (onePost) {
        return onePost
      } else {
        throw new Error("Post not found")
      }
    } catch (err) {
      throw new Error(err)
    }
  }
}

const Mutation = {
  addPost: async (parent, post, verifyJWT) => {
    const authenticatedUser = checkAuth(verifyJWT)
    console.log("authenticatedUser", authenticatedUser)
    try {
      console.log("Post", post)

      const postExists = await PostModel.findOne({
        title: post.title
      })
      if (postExists) {
        throw new UserInputError("Post already exists", {
          error: {
            addPost: "This post has already been created"
          }
        })
      }

      const errors = {}

      function postValidator() {
        if (post.title.trim() === "") {
          console.log("post.title is empty")
          errors.title = "Title must not be empty"
        } else if (post.title.trim() !== "") {
          console.log("post.title is not empty")
        } else {
          console.log("post.title else")
        }

        if (post.description.trim() === "") {
          console.log("post.description is empty")
          errors.description = "Description must not be empty"
        } else if (post.description.trim() !== "") {
          console.log("post.description is not empty")
        } else {
          console.log("post.description else")
        }

        if (post.category.trim() === "") {
          console.log("post.category is empty")
          errors.category = "post category must not be empty"
        } else if (post.category.trim() !== "") {
          console.log("post.category is not empty")
        } else {
          console.log("post.category else")
        }

        return {
          errors
        }
      }

      postValidator()

      if (Object.keys(errors).length !== 0) {
        console.log("Invalid")
        throw new UserInputError("Errors", { errors })
      } else if (Object.keys(errors).length === 0) {
        console.log("It is valid")
        console.log("No Errors", Object.keys(errors).length)
      }

      console.log("ERRORS", errors)

      const newPost = await PostModel.create({
        title: post.title,
        description: post.description,
        category: post.category,
        username: authenticatedUser.username,
        createdAt: new Date().toISOString()
      })

      newPost.save()

      console.log("new post here:", newPost)

      return {
        message: "Post successfully created!",
        ...newPost._doc,
        id: newPost._id
      }
    } catch (error) {
      console.log(error)
      throw error
    }
  },
  editPost: async (parent, post, verifyJWT) => {
    const authenticatedUser = checkAuth(verifyJWT)
    console.log("authenticatedUser", authenticatedUser)
    try {
      const postExists = await PostModel.findById(post.postId)
      if (!postExists) {
        throw new UserInputError("A post with that ID could not be found", {
          error: {
            editPost: "A post with that ID could not be found"
          }
        })
      }

      // Need .save()?

      let updatedPost
      if (authenticatedUser.username === postExists.username) {
        updatedPost = await PostModel.findByIdAndUpdate(post.postId, {
          title: post.title,
          description: post.description,
          category: post.category
        })
      } else {
        throw new AuthenticationError(
          "You do not have permission to edit posts by another user",
          {
            error: {
              editPost:
                "You do not have permission to edit posts by another user"
            }
          }
        )
      }

      console.log("the post", post)

      return {
        message: "Post successfully updated!",
        ...updatedPost._doc,
        id: updatedPost._id
      }
    } catch (error) {
      console.log(error)
      throw error
    }
  },
  deletePost: async (parent, post, verifyJWT) => {
    const authenticatedUser = checkAuth(verifyJWT)
    console.log("authenticatedUser", authenticatedUser)
    console.log("post", post)
    try {
      const postExists = await PostModel.findById(post.postId)
      console.log("postExists", postExists)

      if (postExists) {
        console.log("We have postId", post.postId)
      } else if (post === null) {
        return `The post with the ID of ${post.postId} has already been deleted!`
      }

      if (!postExists) {
        throw new UserInputError("A post with that ID could not be found", {
          error: {
            deletePost: "A post with that ID could not be found"
          }
        })
      }

      let deletedPost
      if (authenticatedUser.username === postExists.username) {
        deletedPost = await PostModel.findByIdAndRemove(post.postId)
      } else {
        throw new AuthenticationError(
          "You do not have permission to delete posts by another user",
          {
            error: {
              deletePost:
                "You do not have permission to delete posts by another user"
            }
          }
        )
      }

      return {
        message: "Post successfully deleted!",
        ...deletedPost._doc,
        id: deletedPost._id
      }
    } catch (error) {
      console.log(error)
      throw error
    }
  }
}

export const postResolvers = {
  Query,
  Mutation
}
