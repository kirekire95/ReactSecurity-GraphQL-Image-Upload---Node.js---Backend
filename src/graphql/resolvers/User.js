import { UserInputError, AuthenticationError } from "apollo-server-express"

import { UserModel } from "../../models/User.js"
import checkAuth from "../../helper/context/index.js"
import {
  createToken,
  hashPassword,
  jwtDecodeToken,
  verifyPassword
} from "../../utilities/utilities.js"

const Query = {
  getUsers: async () => {
    try {
      const allUsers = await UserModel.find({})
      return allUsers
    } catch (err) {
      throw new Error(err)
    }
  },
  getUser: async (parent, user, verifyJWT) => {
    const checkAuthContext = checkAuth(verifyJWT)
    console.log("checkAuthContext", checkAuthContext)
    try {
      const oneUser = await UserModel.findById(user.userId)
      if (oneUser) {
        return oneUser
      } else if (!oneUser) {
        throw new Error("User not found")
      }
    } catch (err) {
      throw new Error(err)
    }
  }
}

const Mutation = {
  loginUser: async (parent, user) => {
    try {
      const { username, password } = user
      const foundUser = await UserModel.findOne({ username })

      const errors = {}

      function userValidator() {
        if (!foundUser) {
          errors.username = "A user with that username could not be found!"
        }

        if (username.trim() === "") {
          console.log("user.username is empty")
          errors.username = "Username must not be empty"
        } else if (username.trim() !== "") {
          console.log("user.username is not empty")
        } else {
          console.log("user.username else")
        }

        if (password.trim() === "") {
          console.log("user.password is empty")
          errors.password = "Password must not be empty"
        } else if (password.trim() !== "") {
          console.log("user.password is not empty")
        } else {
          console.log("user.password else")
        }

        return {
          errors
        }
      }

      userValidator()

      // Must validate and throw errors here first because otherwise it cannot execute isValidPassword properly as it's empty

      console.log("User", user)

      console.log("ERRORS", errors)

      if (Object.keys(errors).length !== 0) {
        console.log("Invalid")
        throw new UserInputError("Errors", { errors })
      } else if (Object.keys(errors).length === 0) {
        console.log("It is valid")
        console.log("No Errors", Object.keys(errors).length)
      }

      const isValidPassword = await verifyPassword(password, foundUser.password)

      if (!isValidPassword && password !== "") {
        errors.password = "Wrong credentials"
        throw new UserInputError("Wrong credentials", { errors })
      } else if (isValidPassword) {
        const userInfo = foundUser

        console.log("Login Userinfo", userInfo)

        const token = createToken(userInfo)
        const expiresAt = jwtDecodeToken(token).expiresAt

        return {
          message: "Authentication successful",
          userInfo,
          errors,
          token,
          expiresAt
        }
      }
    } catch (error) {
      console.log(error)
      throw error
    }
  },
  addUser: async (parent, user) => {
    try {
      const { username, email, password, confirmPassword } = user

      const usernameExists = await UserModel.findOne({ username })
      const emailExists = await UserModel.findOne({ email })

      const errors = {}

      function registerValidator() {
        if (usernameExists) {
          errors.username = "This username has already been taken"
        }

        if (emailExists) {
          errors.email = "This email has already been taken"
        }

        if (username.trim() === "") {
          console.log("user.username is empty")
          errors.username = "Username must not be empty"
        } else if (username.trim() !== "") {
          console.log("user.username is not empty")
        } else {
          console.log("user.username else")
        }

        if (password === "") {
          console.log("user.password is empty")
          errors.password = "Password must not be empty"
        } else if (password.trim() !== "") {
          console.log("user.password is not empty")
        } else {
          console.log("user.password else")
        }

        if (password !== confirmPassword) {
          errors.confirmPassword = "Passwords must match"
        }

        if (email.trim() === "") {
          errors.email = "Email must not be empty"
        } else {
          const regEx = /^([0-9a-zA-Z]([-.\w]*[0-9a-zA-Z])*@([0-9a-zA-Z][-\w]*[0-9a-zA-Z]\.)+[a-zA-Z]{2,9})$/

          if (!email.match(regEx)) {
            errors.email = "Email must be a valid email address"
          }
        }
      }

      registerValidator()

      if (Object.keys(errors).length > 0) {
        throw new UserInputError("Errors", { errors })
      } else {
        console.log("No Errors", Object.keys(errors).length)
      }

      console.log("ERRORS", errors)

      console.log("The User", user)

      const hashedPassword = await hashPassword(password)

      const newUser = await new UserModel({
        username: username.toLowerCase().trim(),
        password: hashedPassword,
        email: email.toLowerCase().trim(),
        createdAt: new Date().toISOString(),
        role: "admin"
      })

      const savedUser = await newUser.save()

      if (savedUser) {
        console.log("SAVEDUSESEERRERRERE", savedUser)
        const token = createToken(savedUser)
        const expiresAt = jwtDecodeToken(token).expiresAt

        console.log("savedUser here:", savedUser)

        const { _id, username, password, email, createdAt, role } = savedUser

        const userInfo = {
          _id,
          username,
          password,
          email,
          createdAt,
          role
        }

        return {
          message: "Account successfully created",
          userInfo,
          errors,
          token,
          expiresAt,
          ...newUser._doc
        }
      } else {
        console.log("Something went wrong creating the account")
      }
    } catch (error) {
      console.log(error)
      throw error
    }
  },
  editUser: async (parent, user, verifyJWT) => {
    const authenticatedUser = checkAuth(verifyJWT)
    console.log("authenticatedUser", authenticatedUser)
    try {
      const { userId, username, password, email } = user
      const userExists = await UserModel.findById(userId)
      if (!userExists) {
        throw new UserInputError("A user with that ID could not be found", {
          error: {
            editUser: "A user with that ID could not be found"
          }
        })
      }

      const hashedPassword = await hashPassword(password)

      let updatedUser
      if (
        authenticatedUser.username === userExists.username ||
        authenticatedUser.username === "kirekire95"
      ) {
        updatedUser = await UserModel.findByIdAndUpdate(userId, {
          username: username,
          password: hashedPassword,
          email: email
        })
      } else {
        throw new AuthenticationError(
          "You do not have permission to edit another user",
          {
            error: {
              editUser: "You do not have permission to edit another user"
            }
          }
        )
      }

      // Need .save()?
      // Need to log out and give new token here?...

      return {
        message: "User edited successfully",
        ...updatedUser._doc,
        id: updatedUser._id
      }
    } catch (error) {
      console.log(error)
      throw error
    }
  },
  deleteUser: async (parent, user, verifyJWT) => {
    const authenticatedUser = checkAuth(verifyJWT)
    console.log("authenticatedUser", authenticatedUser)
    try {
      const { userId } = user
      const userExists = await UserModel.findById(userId)
      console.log("userExists", userExists)

      if (!userExists) {
        throw new UserInputError("A user with that ID could not be found", {
          error: {
            deleteUser: "A user with that ID could not be found"
          }
        })
      }

      let deletedUser

      if (
        authenticatedUser.username === userExists.username ||
        authenticatedUser.username === "kirekire95"
      ) {
        deletedUser = await UserModel.findByIdAndRemove(userId)
      } else {
        throw new AuthenticationError(
          "You do not have permission to delete another user",
          {
            error: {
              deleteUser: "You do not have permission to delete another user"
            }
          }
        )
      }

      return {
        message: "User deleted successfully",
        ...deletedUser._doc,
        id: deletedUser._id
      }
    } catch (error) {
      console.log(error)
      throw error
    }
  }
}

export const userResolvers = {
  Query,
  Mutation
}
