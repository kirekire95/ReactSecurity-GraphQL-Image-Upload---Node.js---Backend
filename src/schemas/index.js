import { gql } from "apollo-server-express"

export const typeDefs = gql`
  type User {
    _id: ID!
    username: String!
    password: String!
    confirmPassword: String!
    email: String!
    createdAt: String!
  }

  type Query {
    getUsers: [User]!
    getUser(userId: ID!): User
  }

  type AuthenticationResult {
    message: String
    userInfo: User!
    token: String!
    expiresAt: String!
  }

  type Mutation {
    loginUser(username: String!, password: String!): AuthenticationResult
    addUser(
      username: String!
      password: String!
      confirmPassword: String!
      email: String!
    ): AuthenticationResult
    editUser(
      userId: ID!
      username: String
      password: String
      confirmPassword: String
      email: String
    ): User
    deleteUser(userId: ID!): User
  }
`
