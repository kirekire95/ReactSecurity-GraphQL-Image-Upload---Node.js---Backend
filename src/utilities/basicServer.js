import { ApolloServer, gql } from "apollo-server-express"

export const createSimpleApolloServer = () => {
  const typeDefs = gql`
    type Book {
      isbn: ID
      name: String
    }
    type Query {
      books: [Book]
    }
  `

  const books = [{ isbn: "0252166276", name: "Harry Potter" }]

  const resolvers = {
    Query: {
      books: () => books
    }
  }

  const server = new ApolloServer({ typeDefs, resolvers })

  return server
}
