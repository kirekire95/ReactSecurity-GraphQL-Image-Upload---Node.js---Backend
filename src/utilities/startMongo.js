import { config } from "dotenv"
config({ path: ".env.local" })
import mongoose from "mongoose"

const { MONGODB } = process.env

export async function startMongoDB() {
  await mongoose.connect(MONGODB, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false
  })
  mongoose.connection.once("open", () =>
    console.log(chalk.hex("#897DDC")("Connected to MongoDB"))
  )
  mongoose.connection.on("error", (error) => console.error(error))
}
