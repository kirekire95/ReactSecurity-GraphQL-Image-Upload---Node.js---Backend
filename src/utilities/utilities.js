import jwt from "jsonwebtoken"
import bcrypt from "bcrypt"
import jwtDecode from "jwt-decode"

const createToken = (user) => {
  console.log("createToken User", user)
  return jwt.sign(
    {
      username: user.username,
      email: user.email,
      role: user.role
    },
    process.env.SECRET,
    {
      expiresIn: "1h",
      algorithm: "HS256"
    }
  )
}

const hashPassword = (password) => {
  return new Promise((resolve, reject) => {
    // Generate a salt at level 12 strength
    bcrypt.genSalt(12, (err, salt) => {
      if (err) {
        reject(err)
      }
      bcrypt.hash(password, salt, (err, hash) => {
        if (err) {
          reject(err)
        }
        resolve(hash)
      })
    })
  })
}

const verifyPassword = (passwordAttempt, hashedPassword) => {
  return bcrypt.compare(passwordAttempt, hashedPassword)
}

const jwtDecodeToken = (passedInToken) => {
  const decodedToken = jwtDecode(passedInToken)
  console.log("jwtDecodeToken token", decodedToken)
  const expiresAt = decodedToken.exp

  return {
    decodedToken,
    expiresAt
  }
}

export { createToken, hashPassword, verifyPassword, jwtDecodeToken }
