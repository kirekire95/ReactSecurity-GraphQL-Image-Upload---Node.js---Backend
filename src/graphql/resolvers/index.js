import { userResolvers } from "./User.js"

export { userResolvers }
