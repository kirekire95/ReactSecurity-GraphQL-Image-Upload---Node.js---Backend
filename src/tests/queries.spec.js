import { createTestClient } from "apollo-server-testing"
import { ApolloServer, gql } from "apollo-server-express"

import { postResolvers, userResolvers } from "../graphql/resolvers"
import { typeDefs } from "../schemas"
import { startMongoDB } from "../utilities/startMongo"
import createMockDB from "../utilities/startMockMongo"

const GET_USERS = gql`
  query getUsers {
    getUsers {
      _id
      username
      password
      email
      createdAt
    }
  }
`

const constructTestServer = () => {
  const server = new ApolloServer({
    typeDefs,
    resolvers: [userResolvers, postResolvers]
  })

  return { server }
}

// describe("Queries", () => {
//   it("fetches an array of users", async () => {
//     const { server } = constructTestServer()

//     const { query } = createTestClient(server)

//     const response = await query({
//       query: GET_USERS
//     })

//     console.log("RESPONSE", response)

//     // expect(response).toMatchSnapshot()
//     expect(response.data.getUsers).toEqual([{ name: "The Colour Of Magic" }])
//   })
// })

const { server } = constructTestServer()

const { query } = createTestClient(server)

describe("User resolver", () => {
  beforeAll(async () => {
    await createMockDB()
  })

  test("get users", async () => {
    const GET_USERS = gql`
      query getUsers {
        getUsers {
          _id
          username
          password
          email
          createdAt
        }
      }
    `

    const {
      data: { getUsers }
    } = await query({ query: GET_USERS })
    expect(getUsers).toEqual([
      {
        id: "5e7df7fc6553acd3ec50fe8f",
        name: "John",
        address: "Privatgatan 15"
      }
    ])
    expect(getUsers).toMatchSnapshot()
  })
})
