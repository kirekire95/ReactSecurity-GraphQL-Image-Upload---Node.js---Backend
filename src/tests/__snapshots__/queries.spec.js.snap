// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries fetches an array of users 1`] = `Promise {}`;
